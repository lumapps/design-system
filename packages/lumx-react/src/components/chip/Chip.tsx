import { Color, ColorPalette, Size, Theme } from '@lumx/react';
import { useStopPropagation } from '@lumx/react/hooks/useStopPropagation';

import { Comp, GenericProps, getRootClassName, handleBasicClasses, onEnterPressed } from '@lumx/react/utils';

import classNames from 'classnames';

import isFunction from 'lodash/isFunction';
import React, { forwardRef, MouseEventHandler, ReactNode } from 'react';

/**
 * Chip sizes.
 */
type ChipSize = Extract<Size, 's' | 'm'>;

/**
 * Defines the props of the component.
 */
export interface ChipProps extends GenericProps {
    /** A component to be rendered after the content. */
    after?: ReactNode;
    /** A component to be rendered before the content. */
    before?: ReactNode;
    /** Color variant. */
    color?: Color;
    /** Whether the component is clickable or not. */
    isClickable?: boolean;
    /** Whether the component is disabled or not. */
    isDisabled?: boolean;
    /** Whether the chip is currently in a highlighted state or not. */
    isHighlighted?: boolean;
    /** Whether the component is selected or not. */
    isSelected?: boolean;
    /** Size variant. */
    size?: ChipSize;
    /** Theme adapting the component to light or dark background. */
    theme?: Theme;
    /** On "after" element clicked callback. */
    onAfterClick?: MouseEventHandler;
    /** On "before" element clicked callback. */
    onBeforeClick?: MouseEventHandler;
}

/**
 * Component display name.
 */
const COMPONENT_NAME = 'Chip';

/**
 * Component default class name and class prefix.
 */
const CLASSNAME = getRootClassName(COMPONENT_NAME);

/**
 * Component default props.
 */
const DEFAULT_PROPS: Partial<ChipProps> = {
    size: Size.m,
    theme: Theme.light,
};

/**
 * Chip component.
 *
 * @param  props Component props.
 * @param  ref   Component ref.
 * @return React element.
 */
export const Chip: Comp<ChipProps, HTMLAnchorElement> = forwardRef((props, ref) => {
    const {
        after,
        before,
        children,
        className,
        color,
        disabled,
        isClickable,
        isDisabled = disabled,
        isHighlighted,
        isSelected,
        onAfterClick,
        onBeforeClick,
        onClick,
        size,
        theme,
        ...forwardedProps
    } = props;
    const hasAfterClick = isFunction(onAfterClick);
    const hasBeforeClick = isFunction(onBeforeClick);
    const hasOnClick = isFunction(onClick);

    // Adapt color to the theme.
    const chipColor = color || (theme === Theme.light ? ColorPalette.dark : ColorPalette.light);

    const handleOnBeforeClick = useStopPropagation(onBeforeClick);
    const handleOnAfterClick = useStopPropagation(onAfterClick);

    return (
        // eslint-disable-next-line jsx-a11y/no-static-element-interactions
        <a
            {...forwardedProps}
            ref={ref}
            className={classNames(
                className,
                handleBasicClasses({
                    clickable: Boolean(hasOnClick) || isClickable,
                    color: chipColor,
                    isDisabled,
                    hasAfter: Boolean(after),
                    hasBefore: Boolean(before),
                    highlighted: Boolean(isHighlighted),
                    prefix: CLASSNAME,
                    selected: Boolean(isSelected),
                    size,
                    unselected: Boolean(!isSelected),
                }),
            )}
            role={hasOnClick ? 'button' : undefined}
            tabIndex={isDisabled || !hasOnClick ? -1 : 0}
            aria-disabled={(hasOnClick && isDisabled) || undefined}
            onClick={hasOnClick ? onClick : undefined}
            onKeyDown={hasOnClick ? onEnterPressed(onClick) : undefined}
        >
            {before && (
                // eslint-disable-next-line jsx-a11y/click-events-have-key-events,jsx-a11y/no-static-element-interactions
                <div
                    className={classNames(`${CLASSNAME}__before`, {
                        [`${CLASSNAME}__before--is-clickable`]: hasBeforeClick,
                    })}
                    onClick={handleOnBeforeClick}
                >
                    {before}
                </div>
            )}

            <div className={`${CLASSNAME}__label`}>{children}</div>

            {after && (
                // eslint-disable-next-line jsx-a11y/click-events-have-key-events,jsx-a11y/no-static-element-interactions
                <div
                    className={classNames(`${CLASSNAME}__after`, {
                        [`${CLASSNAME}__after--is-clickable`]: hasAfterClick,
                    })}
                    onClick={handleOnAfterClick}
                >
                    {after}
                </div>
            )}
        </a>
    );
});
Chip.displayName = COMPONENT_NAME;
Chip.className = CLASSNAME;
Chip.defaultProps = DEFAULT_PROPS;
