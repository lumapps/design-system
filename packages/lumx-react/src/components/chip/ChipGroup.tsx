import { Alignment, HorizontalAlignment } from '@lumx/react/components';
import React, { forwardRef, ReactNode } from 'react';

import classNames from 'classnames';

import { Comp, GenericProps, getRootClassName, handleBasicClasses } from '@lumx/react/utils';

import { useChipGroupNavigation } from '@lumx/react/hooks/useChipGroupNavigation';

/**
 * Defines the props of the component.
 */
export interface ChipGroupProps extends GenericProps {
    /** Chip horizontal alignment. */
    align?: HorizontalAlignment;
    /** List of Chip. */
    children: ReactNode;
}

/**
 * Component default props.
 */
const DEFAULT_PROPS: Partial<ChipGroupProps> = {
    align: Alignment.left,
};

/**
 * Component display name.
 */
const COMPONENT_NAME = 'ChipGroup';

/**
 * Component default class name and class prefix.
 */
const CLASSNAME = getRootClassName(COMPONENT_NAME);

/**
 * ChipGroup component.
 *
 * @param  props Component props.
 * @param  ref   Component ref.
 * @return React element.
 */
const InternalChipGroup: Comp<ChipGroupProps, HTMLDivElement> = forwardRef((props, ref) => {
    const { align, children, className, ...forwardedProps } = props;
    const chipGroupClassName = handleBasicClasses({
        align,
        prefix: CLASSNAME,
    });

    return (
        <div ref={ref} {...forwardedProps} className={classNames(className, chipGroupClassName)}>
            {children}
        </div>
    );
});
InternalChipGroup.displayName = COMPONENT_NAME;
InternalChipGroup.className = CLASSNAME;
InternalChipGroup.defaultProps = DEFAULT_PROPS;

export const ChipGroup = Object.assign(InternalChipGroup, { useChipGroupNavigation });
