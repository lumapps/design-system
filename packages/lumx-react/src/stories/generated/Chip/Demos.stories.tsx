/**
 * File generated when storybook is started. Do not edit directly!
 */
export default { title: 'LumX components/chip/Demos' };

export { App as Choice } from './choice';
export { App as Default } from './default';
export { App as Dismissible } from './dismissible';
export { App as Filter } from './filter';
export { App as Group } from './group';
export { App as Small } from './small';
