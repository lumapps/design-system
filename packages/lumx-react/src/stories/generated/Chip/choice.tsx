/home/gcornut/git/dup-design-system/packages/site-demo/content/product/components/chip/react/choice.tsx