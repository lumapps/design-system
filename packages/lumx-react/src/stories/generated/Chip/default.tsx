/home/gcornut/git/design-system/packages/site-demo/content/product/components/chip/react/default.tsx