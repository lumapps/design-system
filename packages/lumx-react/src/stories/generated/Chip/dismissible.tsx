../../../../../site-demo/content/product/components/chip/react/dismissible.tsx