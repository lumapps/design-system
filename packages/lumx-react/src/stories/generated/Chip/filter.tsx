../../../../../site-demo/content/product/components/chip/react/filter.tsx