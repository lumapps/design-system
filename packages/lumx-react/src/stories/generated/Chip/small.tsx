../../../../../site-demo/content/product/components/chip/react/small.tsx