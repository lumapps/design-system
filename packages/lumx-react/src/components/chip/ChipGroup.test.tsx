import { shallow } from 'enzyme';
import 'jest-enzyme';
import React from 'react';

import { Chip } from './Chip';
import { ChipGroup, ChipGroupProps } from './ChipGroup';

/**
 * Mounts the component and returns common DOM elements / data needed in multiple tests further down.
 */
const setup = (propOverrides: Partial<ChipGroupProps> = {}) => {
    const props = {
        children: [<Chip key="1">Chip 1</Chip>, <Chip key="2">Chip 2</Chip>, <Chip key="3">Chip 3</Chip>],
        ...propOverrides,
    };
    const wrapper = shallow(<ChipGroup {...props} />);

    return { props, wrapper };
};

describe('<ChipGroup />', () => {
    // 1. Test render via snapshot (default state of component).
    describe('Snapshot', () => {
        it('should render correctly Chip Group component', () => {
            const { wrapper } = setup();
            expect(wrapper).toMatchSnapshot();
        });
    });
});
