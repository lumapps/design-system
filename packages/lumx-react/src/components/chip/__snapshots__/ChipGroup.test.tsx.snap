// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<ChipGroup /> Snapshot should render correctly Chip Group component 1`] = `
<div
  className="lumx-chip-group lumx-chip-group--align-left"
>
  <Chip
    key="1"
    size="m"
    theme="light"
  >
    Chip 1
  </Chip>
  <Chip
    key="2"
    size="m"
    theme="light"
  >
    Chip 2
  </Chip>
  <Chip
    key="3"
    size="m"
    theme="light"
  >
    Chip 3
  </Chip>
</div>
`;
